api/valor_api/backend/metrics/classification.py

```diff
@@ -981,26 +1154,173 @@ def _compute_confusion_matrix_and_metrics_at_grouper_key(
+    label_map: LabelMapType, optional
+        Optional mapping of individual labels to a grouper label. Useful when you need to evaluate performance using labels that differ across datasets and models.
+
+    Returns
+    ----------
+    list[ConfusionMatrix, Metric]
+        A list of confusion matrices and metrics.
+    """
+    labels = core.fetch_union_of_labels(
+        db=db,
+        lhs=groundtruth_filter,
+        rhs=prediction_filter,
+    )
+
+    label_mapping = create_label_mapping(
+        db=db,
+        labels=labels,
+        label_map=label_map,
+    )
+
+    groundtruths_subquery = generate_select(
+        models.Datum.id.label("datum_id"),
+        models.Datum.uid.label("datum_uid"),
+        models.Dataset.name.label("dataset_name"),
+        models.Label.id,
+        label_mapping,
+        filters=groundtruth_filter,
+        label_source=models.GroundTruth,
+    ).subquery()
+    groundtruths_cte = (
+        select(
+            groundtruths_subquery.c.datum_id,
+            groundtruths_subquery.c.datum_uid,
+            groundtruths_subquery.c.dataset_name,
+            models.Label.id.label("label_id"),
+            models.Label.key,
+            models.Label.value,
+        )
+        .select_from(groundtruths_subquery)
+        .join(
+            models.Label,
+            models.Label.id == groundtruths_subquery.c.label_id,
+        )
+        .distinct()
+        .cte()
+    )
+
+    predictions_subquery = generate_select(
+        models.Datum.id.label("datum_id"),
+        models.Datum.uid.label("datum_uid"),
+        models.Dataset.name.label("dataset_name"),
+        models.Model.name.label("model_name"),
+        models.Prediction.id.label("prediction_id"),
+        models.Prediction.score,
+        models.Label.id,
+        label_mapping,
+        filters=prediction_filter,
+        label_source=models.Prediction,
+    ).subquery()
+    predictions_cte = (
+        select(
+            predictions_subquery.c.dataset_name,
+            predictions_subquery.c.model_name,
+            predictions_subquery.c.datum_id,
+            predictions_subquery.c.datum_uid,
+            models.Label.id.label("label_id"),
+            models.Label.key,
+            models.Label.value,
+            func.min(predictions_subquery.c.prediction_id).label(
+                "prediction_id"
+            ),
+            func.max(predictions_subquery.c.score).label("score"),
+        )
+        .select_from(predictions_subquery)
+        .join(
+            models.Label,
+            models.Label.id == predictions_subquery.c.label_id,
+        )
+        .group_by(
+            predictions_subquery.c.dataset_name,
+            predictions_subquery.c.model_name,
+            predictions_subquery.c.datum_id,
+            predictions_subquery.c.datum_uid,
+            models.Label.id,
+            models.Label.key,
+            models.Label.value,
+        )
+        .cte()
+    )
+
+    def groundtruth_label_query():
```

Comment:
move these functions out or to the top, make private, add docstring?

Reply:
This is remanent of my profiling, they are now variables!